### PR TITLE
cmd/cue: detect kubernetes semantics for `get go`

### PR DESCRIPTION
**Pull request by fionera**
Kubernetes types with either +optional or omitempty, are expected to not have a null disjunction. By checking for their annotations we can switch to a kubernetes semantic mode and not generate them.

### CODE REVIEW

File: cmd/cue/cmd/get_go.go [hunk 1]
```diff
@@ -444,9 +445,39 @@ func (e *extractor) recordTypeInfo(p *packages.Package) {
 	}
 }
 
+type importSematic int
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go]: given that we only need "default" vs "k8s" for now, I think just a boolean is fine. we can always expand to an enum type in the future if needed. I suspect/hope that we won't need more modes anytime soon.

File: cmd/cue/cmd/get_go.go [hunk 2]
```diff
@@ -444,9 +445,39 @@ func (e *extractor) recordTypeInfo(p *packages.Package) {
...
 func (e *extractor) extractPkg(root string, p *packages.Package) error {
 	e.pkg = p
-	e.logf("--- Package %s", p.PkgPath)
+	e.semantic = semanticDefault
+
+	for _, f := range p.Syntax {
+		for _, c := range f.Comments {
+			for _, c := range c.List {
+				switch c.Text {
+				case "// +k8s:openapi-gen=true":
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go]: personally I might do a `strings.Contains(c.Text(), "+k8s:openapi-gen=true")` - it's simpler, it's less likely to cause weird false negatives with whitespace, and I doubt it would cause any false positives ever.

File: cmd/cue/cmd/get_go.go [hunk 3]
```diff
@@ -1209,10 +1240,13 @@ func (e *extractor) makeType(typ types.Type) (result cueast.Expr) {
 
 		return result
 	case *types.Pointer:
+		if e.semantic == semanticKubernetes && kind == optional {
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go]: make a comment here, like "in k8s, a field is only nullable if explicitly marked as such"

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 4]
```diff
@@ -0,0 +1,601 @@
+# Test that a get go works for kubernetes types,
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:1]: this test file copies all of get_go_types.txtar, which works, but it's very verbose. we only really care about a few edge cases which behave differently in k8s mode, so I'd strongly suggest writing a much smaller test that's focused on that, in 50 lines or so.

**Review by mvdan — COMMENTED** [review]
SGTM overall, but I think this implementation isn't quite complete - it should support the "optional" and "nullable" field comment attributes that the k8s openapi tool supports. Note that we already support `+optional` (and I'm not sure if that's already compatible with what k8s does), but we don't support `+nullable` at all.

**Comment by mvdan** [comment]
Also, could you split this in two commits as follows:
1) Add the testscript file, showing what the current implementation generates for k8s types
2) Add the code changes, showing how it affects the generated types from the first commit, effectively showing a diff removing `null | ` etc

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# including those special cases that result in _ and string
+# that are specifically mentioned in help get go.
+
+# cue get go
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar]: nit: remove this line

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+exec cue get go --local ./...
+cmp ./file2_go_gen.cue ./file2_go_gen.cue.golden
+
+-- go.mod --
+module mod.test
+
+go 1.21
+-- cue.mod/module.cue --
+module: "mod.test"
+-- file1.go --
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:14]: call this `doc.go` to be more consistent with upstream?

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+package pkg1
+
+// +k8s:openapi-gen=true
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:17]: I seem to recall that the files we looked at had this +k8s comment _above the package declaration_. This is important, because it means we don't need to scan the comments from the entire file to see if the comment exists or not. I'd move it up here to be consistent with that expectation.

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+type Foo struct {
+	Required string `json:"required"`
+	Pointer *string `json:"pointer"`
+	OmitEmptyTagPointer *string `json:"omitEmptyTagPointer,omitempty"`
+	// +optional
+    OptionalCommentTagPointer *string `json:"optionalCommentTagPointer"`
+	// +nullable
+    NullableCommentTagPointer *string `json:"nullableCommentTagPointer"`
+}
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:31]: I'd add something like an `+optional` non-pointer string, for completeness.

Also, is the word "Tag" in the fields above adding anything? I'd personally remove it.

File: cmd/cue/cmd/get_go.go [hunk 9]
```diff
@@ -446,7 +447,16 @@ func (e *extractor) recordTypeInfo(p *packages.Package) {
 
 func (e *extractor) extractPkg(root string, p *packages.Package) error {
 	e.pkg = p
-	e.logf("--- Package %s", p.PkgPath)
+	e.k8sSemantic = false
+
+	for _, f := range p.Syntax {
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go:453]: the example from upstream we looked at had the comment in `doc.go`, is that something that we can rely on so that we don't scan all files?

File: cmd/cue/cmd/get_go.go [hunk 10]
```diff
@@ -446,7 +447,16 @@ func (e *extractor) recordTypeInfo(p *packages.Package) {
...
+		for _, c := range f.Comments {
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go:454]: you should only scan up to the package declaration line. not just for the sake of performance, but also so that any comments elsewhere in the file don't incorrectly affect our logic here.

basically: loop over f.Comments, but break as soon as `c.Pos() > f.Package`.

File: cmd/cue/cmd/get_go.go [hunk 11]
```diff
@@ -1480,24 +1491,36 @@ func (e *extractor) isInline(tag string) bool {
...
-	if _, ok := f.Type().(*types.Pointer); ok {
-		return true
-	}
-
+func (e *extractor) detectFieldKind(f *types.Var, doc *ast.CommentGroup, tag string) fieldKind {
+	// From k8s docs:
+	// - a map field marked with `+nullable` would accept either `foo: null` or `foo: {}`.
+	// - Using the `+optional` or the `omitempty` tag means that the field is optional.
+	//   Note that for backward compatibility, any field that has the `omitempty` struct
+	//   tag, and is not explicitly marked as `+required`, will be considered to be optional.
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go:1501]: this doesn't have to happen in this PR, but I'd leave a TODO about supporting `+required`, as it seems to me like we should support it for complete compatibility of these rules here.

File: cmd/cue/cmd/get_go.go [hunk 12]
```diff
@@ -1480,24 +1491,36 @@ func (e *extractor) isInline(tag string) bool {
...
 	for line := range strings.SplitSeq(doc.Text(), "\n") {
 		before, _, _ := strings.Cut(strings.TrimSpace(line), "=")
-		if before == "+optional" {
-			return true
+		switch before {
+		case "+nullable":
+			return regular
+		case "+optional":
+			return optional
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go:1508]: I see from upstream examples that they support key-values like `+featureGate=CRDValidationRatcheting`; do they support e.g. `+optional=false`?

**Review by mvdan — APPROVED** [review]
LGTM - just a few last nits and I'm happy to import into Gerrit and merge

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar (hunk 7 above)
Comment (fionera) [reply cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:17]: Sounds reasonable, tho I found a few things where they are on the struct itself
```
// +k8s:deepcopy-gen=package

package apps
```

```
// +protobuf=true
// +protobuf.options.(gogoproto.goproto_stringer)=false
// +k8s:openapi-gen=true
// +k8s:openapi-model-package=io.k8s.apimachinery.pkg.util.intstr
type IntOrString struct {
	Type   Type   `protobuf:"varint,1,opt,name=type,casttype=Type"`
	IntVal int32  `protobuf:"varint,2,opt,name=intVal"`
	StrVal string `protobuf:"bytes,3,opt,name=strVal"`
}
```

Comment (fionera) [reply cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:17]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go

File: cmd/cue/cmd/get_go.go (hunk 9 above)
Comment (fionera) [reply cmd/cue/cmd/get_go.go:453]: Sadly no, see comment above

File: cmd/cue/cmd/get_go.go (hunk 10 above)
Comment (fionera) [reply cmd/cue/cmd/get_go.go:454]: I think we have to go over all comments, as stated above. I added a break tho.

File: cmd/cue/cmd/get_go.go (hunk 12 above)
Comment (fionera) [reply cmd/cue/cmd/get_go.go:1508]: No, optional doesnt have any arguments

File: cmd/cue/cmd/get_go.go (hunk 11 above)
Comment (fionera) [reply cmd/cue/cmd/get_go.go:1501]: There is a TODO at https://github.com/cue-lang/cue/blob/master/cmd/cue/cmd/get_go.go#L1067

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar (hunk 6 above)
Comment (fionera) [reply cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:14]: Sadly they are not always called doc.go :(

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar (hunk 7 above)
Comment (mvdan) [reply cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:17]: ack, thank you

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# Test that a get go works for kubernetes types,
+# including those special cases that result in _ and string
+# that are specifically mentioned in help get go.
+
+exec cue get go --local ./...
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:5]: nit: replace `./...` with `.`

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+type Foo struct {
+	Required string `json:"required"`
+	Pointer *string `json:"pointer"`
+	OmitEmptyPointer *string `json:"omitEmptyPointer,omitempty"`
+	// +optional
+    OptionalComment string `json:"optionalComment"`
+	// +optional
+    OptionalCommentPointer *string `json:"optionalCommentPointer"`
+	// +nullable
+    NullableCommentPointer *string `json:"nullableCommentPointer"`
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:30]: I just realised - we need to include a field which is both optional and nullable. it should generate as...

```
foo?: null | string
```

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+module: "mod.test"
+-- file1.go --
+package pkg1
+
+// +k8s:openapi-gen=true
+-- file2.go --
+package pkg1
+
+type Foo struct {
+	Required string `json:"required"`
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:22]: For the sake of completeness in this test, I'd rename this to "Regular", and add a "Required" field with the `+required` comment.

File: cmd/cue/cmd/get_go.go [hunk 16]
```diff
@@ -1480,24 +1493,36 @@ func (e *extractor) isInline(tag string) bool {
...
 		}
```
Comment (mvdan) [inline cmd/cue/cmd/get_go.go:1509]: I _think_ that the way this loops over the doc lines, and returns at the first hit, will not work well when a field is marked as BOTH optional and nullable. You might need to make fieldKind a bit set, so that we can track that a field was marked as either or both optional/nullable?

If this is not common and you want to leave this tweak for a future PR, that's fine, but then we'd need a TODO.

**Review by mvdan — APPROVED** [review]
Thanks! LGTM with some last nits.

When you're ready, could you please send these for a final review and test/merge on Gerrit? The way we import PRs to Gerrit doesn't really support multi-commit PRs unless we squash them, and I don't want to squash here.

File: cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar (hunk 14 above)
Comment (fionera) [reply cmd/cue/cmd/testdata/script/get_go_types_kubernetes.txtar:30]: Turns out that should be the default behavior:

```
Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
They are marked with the omitempty json struct tag in the Go definition.
The API server should allow POSTing and PUTing a resource with this field unset.

```

File: cmd/cue/cmd/get_go.go (hunk 16 above)
Comment (fionera) [reply cmd/cue/cmd/get_go.go:1509]: I actually changed the whole PR :sweat_smile: I am gonna close this one and push it to gerrit as its probably a better review platform
